// @nuxt/content@3.12.0/dist/features/llms/runtime/server/utils.js
import { withBase } from "ufo";
import { pascalCase } from "scule";
import { toHast } from "minimark/hast";
import manifest from "#content/manifest";
const linkProps = ["href", "src", "to"];
export async function createDocumentGenerator() {
  const { visit } = await import("unist-util-visit");
  const { stringifyMarkdown } = await import("@nuxtjs/mdc/runtime");
  return generateDocument;
  async function generateDocument(doc, options) {
    const hastTree = refineDocumentBody(doc.body, options);
    let markdown = await stringifyMarkdown(hastTree, {});
    if (!markdown?.trim().startsWith("# ")) {
      const title = doc.title || doc.seo?.title || "";
      if (title) {
        markdown = `# ${title}

${markdown}`;
      }
    }
    return markdown;
  }
  function refineDocumentBody(body, options) {
    const hastTree = toHast(body);
    visit(hastTree, (node) => {
      const el = node;
      if (!el.props?.to && !el.props?.href && !el.props?.src) {
        return;
      }
      for (const prop of linkProps) {
        if (el.props?.[prop]) {
          el.props[prop] = withBase(el.props[prop], options.domain);
        }
      }
    });
    return hastTree;
  }
}
export function prepareContentSections(sections) {
  const contentSections = sections.filter((section) => section.contentCollection);
  if (contentSections.length) {
    return;
  }
  const pageCollecitons = Object.keys(manifest).filter((c) => manifest[c].type === "page");
  const autoGeneratedSections = pageCollecitons.map((c) => ({
    __nuxt_content_auto_generate: true,
    title: pascalCase(c),
    contentCollection: c,
    contentFilters: [
      {
        field: "extension",
        operator: "=",
        value: "md"
      }
    ]
  }));
  sections.push(...autoGeneratedSections);
}
