// nuxt-site-config-kit@3.2.21/dist/shared/nuxt-site-config-kit.C9rm3WhC.cjs
'use strict';

const kit = require('@nuxt/kit');
const urls = require('site-config-stack/urls');
const stdEnv = require('std-env');
const pkgTypes = require('pkg-types');
const siteConfigStack = require('site-config-stack');
const util = require('../util.cjs');

async function initSiteConfig(nuxt = kit.tryUseNuxt()) {
  if (!nuxt)
    return;
  let siteConfig = nuxt._siteConfig;
  if (siteConfig)
    return siteConfig;
  siteConfig = siteConfigStack.createSiteConfigStack();
  const rootDir = nuxt?.options.rootDir || process.cwd?.() || false;
  siteConfig.push({
    _context: "system",
    _priority: -15,
    name: rootDir ? rootDir.split("/").pop() : void 0,
    env: process.env.NODE_ENV
  });
  if (rootDir) {
    const pkgJson = await pkgTypes.readPackageJSON(rootDir);
    if (pkgJson) {
      siteConfig.push({
        _context: "package.json",
        _priority: -10,
        name: pkgJson.name,
        description: pkgJson.description
      });
    }
  }
  siteConfig.push({
    _context: "vendorEnv",
    _priority: -5,
    url: [
      // vercel
      process.env.VERCEL_URL,
      process.env.NUXT_ENV_VERCEL_URL,
      // netlify
      process.env.URL,
      // cloudflare pages
      process.env.CF_PAGES_URL
    ].find((k) => Boolean(k)),
    name: [
      // vercel
      process.env.NUXT_ENV_VERCEL_GIT_REPO_SLUG,
      // netlify
      process.env.SITE_NAME
    ].find((k) => Boolean(k))
  });
  const runtimeConfig = nuxt.options.runtimeConfig;
  const runtimeConfigEnvKeys = [
    ...Object.entries(runtimeConfig.site || {}).filter(([k]) => k.startsWith("site")).map(([k, v]) => [k.replace(/^site/, ""), v]),
    ...Object.entries([...Object.entries(runtimeConfig), ...Object.entries(runtimeConfig.public)]).filter(([k]) => k.startsWith("site")).map(([k, v]) => [k.replace(/^site/, ""), v])
  ];
  siteConfig.push({
    _priority: -2,
    _context: "legacyRuntimeConfig",
    ...Object.fromEntries(runtimeConfigEnvKeys)
  });
  siteConfig.push({
    _context: "buildEnv",
    _priority: -1,
    ...siteConfigStack.envSiteConfig(process.env || {})
  });
  nuxt._siteConfig = siteConfig;
  return siteConfig;
}
async function installNuxtSiteConfig(nuxt = kit.tryUseNuxt()) {
  await kit.installModule(await kit.resolvePath("nuxt-site-config"));
  await initSiteConfig(nuxt);
}
function getSiteConfigStack(nuxt = kit.tryUseNuxt()) {
  if (!nuxt)
    throw new Error("Nuxt context is missing.");
  if (!nuxt._siteConfig)
    throw new Error("Site config is not initialized. Make sure you are running your module after nuxt-site-config.");
  return nuxt._siteConfig;
}
function updateSiteConfig(input, nuxt = kit.tryUseNuxt()) {
  const container = getSiteConfigStack(nuxt);
  return container.push(input);
}
function useSiteConfig(nuxt = kit.tryUseNuxt()) {
  const container = getSiteConfigStack(nuxt);
  return container.get();
}

function withSiteTrailingSlash(path) {
  const siteConfig = useSiteConfig();
  return urls.fixSlashes(siteConfig.trailingSlash, path);
}
function createSitePathResolver(options = {}, nuxt = kit.useNuxt()) {
  const siteConfig = useSiteConfig();
  const nitroOrigin = util.getNitroOrigin();
  const canUseSiteUrl = (options.canonical !== false || stdEnv.env.prerender) && siteConfig.url;
  const nuxtBase = nuxt.options.app.baseURL || "/";
  return (path) => {
    return urls.resolveSitePath(path, {
      ...options,
      siteUrl: canUseSiteUrl ? siteConfig.url : nitroOrigin,
      trailingSlash: siteConfig.trailingSlash,
      base: nuxtBase
    });
  };
}
function withSiteUrl(path, options = {}) {
  const siteConfig = useSiteConfig();
  if (!siteConfig.url && options.throwErrorOnMissingSiteUrl)
    throw new Error("Missing url in site config. Please add `{ site: { url: <url> } }` to nuxt.config.ts.");
  const nuxt = kit.useNuxt();
  const base = nuxt.options.app.baseURL || "/";
  return urls.resolveSitePath(path, {
    absolute: true,
    siteUrl: siteConfig.url || "",
    trailingSlash: siteConfig.trailingSlash,
    base,
    withBase: options.withBase
  });
}

exports.createSitePathResolver = createSitePathResolver;
exports.getSiteConfigStack = getSiteConfigStack;
exports.initSiteConfig = initSiteConfig;
exports.installNuxtSiteConfig = installNuxtSiteConfig;
exports.updateSiteConfig = updateSiteConfig;
exports.useSiteConfig = useSiteConfig;
exports.withSiteTrailingSlash = withSiteTrailingSlash;
exports.withSiteUrl = withSiteUrl;
